Draft-Accepted/RFCXXXX-Command-Shell-Integration.md

```diff
@@ -0,0 +1,287 @@
+but it is recommended to do so.
+
+```json
+{
+    "$schema": "https://schemas.microsoft.com/commandshell/2023/05/01/command.json",
+    "executable": "az",
+    "version": "1.0.0"
+}
+```
+
+The `$schema` includes the version of the manifest with the URL pointing to a published JSON schema.
+The `executable` is the name of the executable that the manifest is for and expected to be found in `$env:PATH`
+The `version` is the version of the manifest for the tool and only used for informational purposes.
+
+### Structured Output and Custom Formatting Registration
+
+Tools can indicate they can output [JSONLines](https://jsonlines.org/) by default and also provide custom formatting.
+
+```json
+{
+    "$schema": "https://schemas.microsoft.com/commandshell/2023/05/01/command.shellintegration.json",
+    "executable": "az",
+    "version": "1.0.0",
+    "structuredOutput": {
+        "default": "jsonlines"
+    }
+}
+```
+
+If a tool indicates it can output structured data in a format supported by the shell,
+the shell will set the env var `COMMAND_SHELL_STRUCTURED_OUTPUT` with the value `jsonlines` in this case
+in the environment block of the process that is launched.
+
+Tools should check for this environment variable and output structured data in the format indicated if supported.
+
+If the tool calls other executables, it is responsible for removing or propagating the environment variable which
+can affect the behavior of child processes.
+
+JSON output should include a `$typeNames` member which is an array of strings that indicate the type hierarchy of the object.
+PowerShell would use this as the same as the `TypeNames` member of a PSObject for formatting.
+Future enhancement would support a JSON defined formatting schema that could be used by other shells.
+
+### Tab Completion Registration
+
+Within the same manifest, a tool can indicate it supports tab completion.
+
+This can either be an executable with arguments.
+
+Arguments can be built-in variables to provide context for the tab completion.
+
+Currently supported variables:
+
+- `{commandLine}` - The full command line that the user has typed so far.
+- `{cursorPosition}` - The current cursor position within the command line, starting at 0 index.
+
+```json
+{
+    "$schema": "https://schemas.microsoft.com/commandshell/2023/05/01/command.shellintegration.json",
+    "executable": "az",
+    "version": "1.0.0",
+    "tabCompletion": {
+        "command": {
+            "executable": "az",
+            "arguments": [
+                "completion",
+                "powershell",
+                "--commandLine",
+                "{commandLine}",
+                "--cursorPosition",
+                "{cursorPosition}"
+            ]
+        }
+    }
+}
+```
+
+or a shell specific script:
+
+```json
+{
+    "$schema": "https://schemas.microsoft.com/commandshell/2023/05/01/command.shellintegration.json",
+    "executable": "az",
+    "version": "1.0.0",
+    "tabCompletion": {
```

Comment:
How can we easily enable 3rd-party completers for a native tool, such as `git`? 3rd-party completers is usually in the form of a PowerShell module, and it needs to register the completer using `Register-ArgumentCompleter` which currently depends on the user put something in their `$PROFILE`.

In this scenario, it doesn't control the shipment of a actual native tool, and requiring the module to deploy a `.command.json` file to a path in `$env:PATH` seems quite unintuitive.

IMHO, I still prefer the prevailing approach in bash/zsh/fish for registering completers. I proposed a similar mechanism in PowerShell in the following 2 comments:
1. https://github.com/PowerShell/PowerShell/issues/17283#issuecomment-1374230415
2. https://github.com/PowerShell/PowerShell/issues/17283#issuecomment-1376515230

For 3rd-party completers that are shipped as PowerShell modules, the module could ideally just have a new field in the `PrivateData` section of the module manifest for completer metadata (e.g. declare a command to run to register the completer), and then `PowerShellGet` can take care of creating the completer file for the module in the completer path when installing the module.

Reply:
With file system based completer registration, it's also easy for a user to override a native tool's default completer.
If we go with `.command.json` for completer registration, how could a user to override or remove a tool's default completer/predictor/feedback provider?

Reply:
I think to override, it would make sense to have "last one wins" since user paths are typically after system paths.  For modules that want similar capability, I think that should be a separate spec and proposal.

Reply:
> For modules that want similar capability, I think that should be a separate spec and proposal.

I think this RFC needs to consider it as well. Native tools may ship PowerShell module along with it, for example, WinGet may choose to ship its Crescendo-based PowerShell module with WinGet, then which mechanism should it follow?

Reply:
Based on above, this would be first-one-wins.  The Crescendo scenario doesn't apply here since there's no conflict.  It's native tab completion vs cmdlets and both can be used together unless I'm misunderstanding.

Reply:
I was just using "_WinGet shipping its Crescendo-based PowerShell module along with its install_" as an example for the scenario that a native tool may choose to ship PowerShell modules along its install. If we completely discard the involvement of PowerShell module, then in that scenario, which mechanism should the native tool follow?

Reply:
I think a whole module would need to fall under PSModulePath and we don't want to do something separate from that.  However, I'll add a section on how modules can leverage extension autoloading.